### ISSUE TITLE
[Discuss] GraphQL::Tracing 

### ISSUE THREAD

**Issue by rmosolgo** [rmosolgo/graphql-ruby#896]
Last year I worked up some monitoring code for graphql-pro, but: 

- Monitoring is required for adopting GraphQL, it's not a nice-to-have; AND
- I think we can do better than graphql-pro's system :) 

I think we should add `GraphQL::Tracing` with an API inspired by `ActiveSupport::Notifications`, for example: 

```ruby 
# For example, tracing the parser 
ast = GraphQL::Tracing.trace("language.parse", { string: query_string }) {
  GraphQL.parse(query_string) 
}

# Or, tracing validation: 
errors = GraphQL::Tracing.trace("language.validate", { document: ast, schema: schema }) {
  schema.validate(ast)
}
```

Subscribers can be added by key: 

```ruby 
GraphQL::Tracing.subscribe("language.parse") do |payload|
  payload[:string] # => " { ... }" 
  payload[:returned] # => #<Language::Nodes::Document> 
  payload[:started_at] # => #<DateTime ...>
  payload[:ended_at] # => #<DateTime ...>
end 

# Or, partial key 
GraphQL::Tracing.subscribe("execution") do { |payload| ... }

# Or, everything 
GraphQL::Tracing.subscribe do { |payload| ... }
```

So, we can implement several "backends" for New Relic, ActiveSupport::Notifications, [Jaeger](http://jaeger.readthedocs.io/en/latest/), etc, which subscribe to these hooks. Apollo Optics could also consume these hooks. 

__Design Goals__: 

- Backend-agnostic 
- Monitor _more_ things (parse, validate, analyze, execute, others?)
- Thread-safe, low-overhead implementation

__Implementation__: 

1. Implement and test `GraphQL::Tracing`
1. Decide naming conventions for keys, add specific traces to the gem source 
1. Implement backends (maybe do this along the way, to make sure we support different services)

What do you think of this design? Anything missing, or anything that could be changed or improved?

(This was also discussed in the [Hack Day Topics thread](https://github.com/emilebosch/graphql-hackathon-ams/issues/1))

**Comment by xuorig** [comment]
😍

**Comment by emilebosch** [comment]
I'm totally in love with this idea. I think its also cool if we can add some metadata, to the current traced object. For instance the SQL query or elastic search query. 

This metadata can then be used to see _what_  is taking that time.

**Comment by rmosolgo** [comment]
Regarding metadata, my thought was that the second argument to `.trace` could be an arbitrary hash: 

```ruby 
GraphQL::Tracing.trace("language.validate", { document: ast, schema: schema })
#                                           ^^^ metadata
```

Then, that hash is yielded to `.subscribe` handlers:

```ruby 
GraphQL::Tracing.subscribe("language.validate") do |payload| 
#                                                  ^^^ metadata 
```

I _think_ that's how AS::Notifications does it. Does that sound like it would do the job?

**Comment by cjoudrey** [comment]
Somewhat related, I was interested in building something inspired by Sangria's slow query log:
https://medium.com/@oleg.ilyenko/profiling-graphql-67330511f09c / https://github.com/sangria-graphql/sangria-slowlog

**tl;dr;** Sangria has a middleware you can use to log queries that take more than X time to execute.

Execution time and resolver time are inlined with the logged query.

**Example**:

```graphql
# [Execution Metrics] duration: 12362ms, validation: 0ms, reducers: 0ms
#
# $id = "1000"
query Test($id: String!) {
  # [Query] count: 1, time: 2ms
  #
  # $id = "1000"
  human(id: $id) {
    # [Human] count: 1, time: 0ms
    name

    # [Human] count: 1, time: 11916ms
    appearsIn

    # [Human] count: 1, time: 358ms
    friends {
      # [Droid] count: 2, min: 0ms, max: 0ms, mean: 0ms, p75: 0ms, p95: 0ms, p99: 0ms
      # [Human] count: 2, min: 0ms, max: 0ms, mean: 0ms, p75: 0ms, p95: 0ms, p99: 0ms
      name
    }
  }
}
```

The middleware can also optionally include these timings as part of the GraphQL result within a new `"extensions": { }` part of the response:

```json
{
  "data": {
    "human": {
      "name": "Luke Skywalker",
      "appearsIn": ["NEWHOPE", "EMPIRE", "JEDI"],
      "friends": [
        {"name": "Han Solo"},
        {"name": "Leia Organa"},
        {"name": "C-3PO"},
        {"name": "R2-D2"}
      ]
    }
  },
  "extensions": {
    "metrics": {
      "executionMs": 362,
      "validationMs": 0,
      "reducersMs": 0,
      "query": "# [Execution Metrics] duration: 362ms, validation: 0ms, reducers: 0ms\n#\n# $id = \"1000\"\nquery Test($id: String!) {\n  # [Query] count: 1, time: 2ms\n  #\n  # $id = \"1000\"\n  human(id: $id) {\n    # [Human] count: 1, time: 0ms\n    name\n\n    # [Human] count: 1, time: 216ms\n    appearsIn\n\n    # [Human] count: 1, time: 358ms\n    friends {\n      # [Droid] count: 2, min: 0ms, max: 0ms, mean: 0ms, p75: 0ms, p95: 0ms, p99: 0ms\n      # [Human] count: 2, min: 0ms, max: 0ms, mean: 0ms, p75: 0ms, p95: 0ms, p99: 0ms\n      name\n    }\n  }\n}",
      "types": {
        "Human": {
          "friends": {
            "count": 1, "minMs": 358, "maxMs": 358, "meanMs": 358,
            "p75Ms": 358, "p95Ms": 358, "p99Ms": 358
          },
          "appearsIn": {
            "count": 1, "minMs": 216, "maxMs": 216, "meanMs": 216,
            "p75Ms": 216, "p95Ms": 216, "p99Ms": 216
          },
          "name": {
            "count": 3, "minMs": 0, "maxMs": 0, "meanMs": 0,
            "p75Ms": 0, "p95Ms": 0, "p99Ms": 0
          }
        },
        "Query": {
          "human": {
            "count": 1, "minMs": 2, "maxMs": 2, "meanMs": 2,
            "p75Ms": 2, "p95Ms": 2, "p99Ms": 2
          }
        },
        "Droid": {
          "name": {
            "count": 2, "minMs": 0, "maxMs": 0, "meanMs": 0,
            "p75Ms": 0, "p95Ms": 0, "p99Ms": 0
          }
        }
      }
    }
  }
}
```

**Comment by emilebosch** [comment]
From that article i love how it looks too
<img width="1045" alt="screen shot 2017-08-17 at 21 58 02" src="https://user-images.githubusercontent.com/303135/29431150-3682dc48-8397-11e7-944d-bc9be9dc4d42.png">
